## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [97]:
# Use the describe function to gather some basic statistics
print(df.describe())

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number      subtotal  \
count  54639.000000  54639.000000  5.463900e+04  54639.000000  5.463900e+04   
mean      99.446073      5.004116  5.702646e+02      2.979667  7.710286e+04   
std      133.164267      5.326599  1.879552e+04    

In [99]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.dtypes

first              object
last               object
job                object
phone              object
email              object
client_id           int64
order_id            int64
order_date         object
order_week          int64
order_year          int64
item_id            object
category           object
subcategory        object
unit_price        float64
unit_cost         float64
unit_weight       float64
qty                 int64
line_number         int64
subtotal          float64
shipping_price    float64
total weight      float64
shipping_rate       int64
total_price       float64
cost              float64
profit            float64
dtype: object

In [101]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   first           54639 non-null  object 
 1   last            54639 non-null  object 
 2   job             54639 non-null  object 
 3   phone           54639 non-null  object 
 4   email           54639 non-null  object 
 5   client_id       54639 non-null  int64  
 6   order_id        54639 non-null  int64  
 7   order_date      54639 non-null  object 
 8   order_week      54639 non-null  int64  
 9   order_year      54639 non-null  int64  
 10  item_id         54639 non-null  object 
 11  category        54639 non-null  object 
 12  subcategory     54639 non-null  object 
 13  unit_price      54639 non-null  float64
 14  unit_cost       54639 non-null  float64
 15  unit_weight     54639 non-null  float64
 16  qty             54639 non-null  int64  
 17  line_number     54639 non-null 

In [102]:
print(df.isnull().sum())

first             0
last              0
job               0
phone             0
email             0
client_id         0
order_id          0
order_date        0
order_week        0
order_year        0
item_id           0
category          0
subcategory       0
unit_price        0
unit_cost         0
unit_weight       0
qty               0
line_number       0
subtotal          0
shipping_price    0
total weight      0
shipping_rate     0
total_price       0
cost              0
profit            0
dtype: int64


In [53]:
# What three item categories had the most entries?
cat_count = df['category'].value_counts()
top_3 = cat_count.head(3)
top_3



category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [54]:
# For the category with the most entries, which subcategory had the most entries?
top_cat = top_3.idxmax()
print(f"Top Category is {top_cat}")

top_cat_df = df[df['category'] == top_cat]

sub_cat_counts = top_cat_df['subcategory'].value_counts()
top_sub_cat = sub_cat_counts.idxmax()

print(f"Top Sub Category is {top_sub_cat}")

Top Category is consumables
Top Sub Category is bathroom supplies


In [55]:
# Store the client ids of those top 5 clients in a list.
client_counts = df['client_id'].value_counts()
top_five_clients = client_counts.head(5)
top_five_clients


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [58]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_id = top_five_client_ids[0]
print(f"Client with the most entries: {top_client_id}")

top_client_df = df[df['client_id'] == top_client_id]
total_units = top_client_df['qty'].sum()
print(f"Total units ordered: {total_units}")

Client with the most entries: 33615
Total units ordered: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [65]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.head()
# df[['first','last','client_id','qty','subtotal']].head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [111]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total weight'] = df['unit_weight'] * df['qty']

def calculate_shipping(total_weight):
    if total_weight > 50:
        return total_weight * 7
    else:
        return total_weight * 10
df['shipping_rate'] = df['total weight'].apply(calculate_shipping)
df.head(10)


first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year  ... unit_weight  qty  \
0   8953482  2023-04-28          17        2023  ...        7.50  105   
1   8069089  2023-05-19          20        2023  ...        1.49   21   
2   1902144  2023-01-29           4        2023  ...        1.68   39   
3   9031802  2023-04-25          17        2023  ...        1.23   29   
4   1322274  2023-05-28          21        2023  ...       46.43   20   
5   1327730  2023-05-02          18        2023  ...        2.79   78   
6   8080280  2023-04-19          16        2023  ...        1.73   56   
7   8946782  2023-02-14           7        2023  ...        5.12   22   
8   9394118  2023-05-13          19        2023  ...        0.27   49   
9   9665852  2023-04-14          15        2023  ...        2.47   78   

  line_number   subtotal  shipping_price  total weight  shipping_rate  \
0           1  115164.00         5512.50        787.50        5512.50   
1           0     523.95          312.90         31.29         312.90   
2           6     527.28          458.64         65.52         458.64   
3           3    1056.18          356.70         35.67         356.70   
4           1    3902.00         6500.20        928.60        6500.20   
5           0   13481.52         1523.34        217.62        1523.34   
6           5   13751.36          678.16         96.88         678.16   
7           4     461.78          788.48        112.64         788.48   
8           3    4109.14          132.30         13.23         132.30   
9           1    5216.64         1348.62        192.66        1348.62   

   total_price       cost    profit  
0    131839.08  120676.50  11162.58  
1       914.26     836.85     77.41  
2      1077.12     985.92     91.20  
3      1543.57    1412.88    130.69  
4     11364.40   10402.20    962.20  
5     16392.81   15004.86   1387.95  
6     15764.25   14429.52   1334.73  
7      1365.91    1250.26    115.65  
8      4633.77    4241.44    392.33  
9      7172.55    6565.26    607.29  

[10 rows x 25 columns]

In [84]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
sales_tax_rate = 0.0925
df['total_price'] = (df['subtotal'] + df['shipping_price']) * (1 + sales_tax_rate)
#round becuase number is crazy
df['total_price'] = df['total_price'].round(2)

df[['subtotal', 'shipping_price', 'total_price']].head()

subtotal  shipping_price  total_price
0  115164.00         5512.50    131839.08
1     523.95          312.90       914.26
2     527.28          458.64      1077.12
3    1056.18          356.70      1543.57
4    3902.00         6500.20     11364.40

In [107]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['cost'] = (df['unit_price'] * df['qty']) + df['shipping_price']

df[['unit_price', 'qty', 'shipping_price', 'cost']].head()

unit_price  qty  shipping_price       cost
0     1096.80  105         5512.50  120676.50
1       24.95   21          312.90     836.85
2       13.52   39          458.64     985.92
3       36.42   29          356.70    1412.88
4      195.10   20         6500.20   10402.20

In [106]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['cost']

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number   subtotal  \
0          17        2023  ...        7.50  105           1  115164.00   
1          20        2023  ...        1.49   21           0     523.95   
2           4        2023  ...        1.68   39           6     527.28   
3          17        2023  ...        1.23   29           3    1056.18   
4          21        2023  ...       46.43   20           1    3902.00   

   shipping_price  total weight  shipping_rate  total_price       cost  \
0         5512.50        787.50              7    131839.08  120676.50   
1          312.90         31.29             10       914.26     836.85   
2          458.64         65.52              7      1077.12     985.92   
3          356.70         35.67             10      1543.57    1412.88   
4         6500.20        928.60              7     11364.40   10402.20   

     profit  
0  11162.58  
1     77.41  
2     91.20  
3    130.69  
4    962.20  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [109]:
# Check your work using the totals above
order_receipts = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

for order_id, expected_total in order_receipts.items():

    order_df = df[df['order_id'] == order_id]
    
    calculated_total = order_df['total_price'].sum()
    
    print(f"Order ID {order_id}:")
    print(f"Calculated Total Price: ${calculated_total:,.2f}")
    print(f"Expected Total Price:   ${expected_total:,.2f}")
    
    if abs(calculated_total - expected_total) == 0:  
        print("PASS")
    else:
        print("FAIL")
   

Order ID 2742071:
Calculated Total Price: $152,811.89
Expected Total Price:   $152,811.89
FAIL
Order ID 2173913:
Calculated Total Price: $162,388.71
Expected Total Price:   $162,388.71
PASS
Order ID 6128929:
Calculated Total Price: $923,441.25
Expected Total Price:   $923,441.25
PASS


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [113]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_five_clients_df = df[df['client_id'].isin(top_five_clients.index)]

# Group by client_id and calculate the total spending
client_spending = top_five_clients_df.groupby('client_id')['total_price'].sum()


print(client_spending)


client_id
24741    82268892.02
33615     8377308.52
38378    12906550.87
46820     9743794.36
66037    10259514.79
Name: total_price, dtype: float64


In [117]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = top_five_clients_df.groupby('client_id').agg(
    total_units_purchased=pd.NamedAgg(column='qty', aggfunc='sum'),
    total_shipping_price=pd.NamedAgg(column='shipping_price', aggfunc='sum'),
    total_revenue=pd.NamedAgg(column='total_price', aggfunc='sum'),
    total_profit=pd.NamedAgg(column='profit', aggfunc='sum')
).reset_index()

summary_df

client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862            5126448.37    82268892.02   
1      33615                  64313            1828984.89     8377308.52   
2      38378                  73667            3429455.40    12906550.87   
3      46820                  75768            1601448.84     9743794.36   
4      66037                  43018            1395151.85    10259514.79   

   total_profit  
0    6965558.40  
1     709291.52  
2    1092774.32  
3     824989.49  
4     868654.56

In [139]:


# Assuming the column names are correct, here’s the updated function
def currency_to_millions(amount):
    return amount / 1_000_000

# Apply the function to the correct columns for total revenue and total profit
summary_df[['Total Revenue (Millions)', 'Total Profit (Millions)']] = summary_df[['Total Revenue (Millions)', 'Total Profit (Millions)']].round(2)

# Rename the columns to indicate values are in millions
summary_df = summary_df.rename(columns={
    'total_revenue': 'Total Revenue (Millions)',
    'total_profit': 'Total Profit (Millions)',
    'total_units_purchased': 'Total Units Purchased',
    'total_shipping_price': 'Total Shipping Price'
})

summary_df

client_id  Total Units Purchased  Total Shipping Price  \
0      24741                 239862            5126448.37   
1      33615                  64313            1828984.89   
2      38378                  73667            3429455.40   
3      46820                  75768            1601448.84   
4      66037                  43018            1395151.85   

   Total Revenue (Millions)  Total Profit (Millions)  
0                     82.27                     6.97  
1                      8.38                     0.71  
2                     12.91                     1.09  
3                      9.74                     0.82  
4                     10.26                     0.87

In [137]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary_df = summary_df.sort_values(by='Total Profit (Millions)', ascending=False).reset_index(drop=True)
sorted_summary_df

client_id  Total Units Purchased  Total Shipping Price  \
0      24741                 239862            5126448.37   
1      38378                  73667            3429455.40   
2      66037                  43018            1395151.85   
3      46820                  75768            1601448.84   
4      33615                  64313            1828984.89   

   Total Revenue (Millions)  Total Profit (Millions)  
0                     82.27                     6.97  
1                     12.91                     1.09  
2                     10.26                     0.87  
3                      9.74                     0.82  
4                      8.38                     0.71